Documentation/Evolution/RegexLiteralPitch.md

```diff
@@ -0,0 +1,257 @@
+- Fit in with Swift's library extensibility story for literals
+
+Further motivation, examples, and discussion can be found in the [overview thread][overview].
+
+## Proposed Solution
+
+We propose the introduction of a regular expression literal that supports [the PCRE syntax][PCRE], in addition to new standard library protocols `ExpressibleByRegexLiteral` and `RegexLiteralProtocol` that allow for the customization of how the regex literal is interpreted (similar to [string interpolation][stringinterpolation]). The compiler will parse the PCRE syntax within a regex literal, and synthesize calls to corresponding builder methods. Types conforming to `ExpressibleByRegexLiteral` will be able to provide a builder type that opts into supporting various regex constructs through the use of normal function declarations and `@available`.
+
+_Note: This pitch concerns language syntax and compiler changes alone, it isn't stating what features the stdlib should support in the initial version or in future versions._
+
+## Detailed Design
+
+A regular expression literal will be introduced using `/` delimiters, within which the compiler will parse [PCRE regex syntax][PCRE]:
+
+```swift
+// Matches "<identifier> = <hexadecimal value>", extracting the identifier and hex number
+let regex = /([[:alpha:]]\w*) = ([0-9A-F]+)/
+```
+
+The above regex literal will be inferred to be the default regex literal type `Regex`. Errors in the regex will be diagnosed by the compiler.
+
+_`Regex` here is a stand-in type, further details about the type such as if or how this will scale to strongly typed captures is still under investigation._
+
+_How best to diagnose grapheme-semantic concerns is still under investigation and probably best discussed in their corresponding threads. For example, `Range<Character>` is not [countable][countable] and [ordering is not linguistically meaningful][ordering], so validating character class ranges may involve restricting to a semantically-meaningful range (e.g. ASCII). This is best discussed in the (upcoming) character class pitch/thread._
+
+The compiler will then transform the literal into a set of builder calls that may be customized by adopting the `ExpressibleByRegexLiteral` protocol. Below is a straw-person transformation of this example: 
+
+```swift
+// let regex = /([[:alpha:]]\w*) = ([0-9A-F]+)/
+let regex = {
+  var builder = T.RegexLiteral()
+
+  // __A4 = /([[:alpha:]]\w*)/
+  let __A1 = builder.posixCharacterClassAlpha()
+  let __A2 = builder.characterClassWord()
+  let __A3 = builder.concatenate(__A1, __A2)
+  let __A4 = builder.captureGroup(__A3)
+
+  // __B1 = / = /
+  let __B1 = builder.literal(" = ")
+
+  // __C3 = /([0-9A-F]+)/
+  let __C1 = builder.customCharacterClass(["0"..."9", "A"..."F"])
+  let __C2 = builder.oneOrMore(__C1)
+  let __C3 = builder.captureGroup(__C2)
+
+  let __D1 = builder.concatenate(__A4, __B1, __C3)
+  return T(regexLiteral: builder.finalize(__D1))
+}()
+```
```

Comment:
This might be premature, but I have Thoughts.

1. We are both calling methods on a mutable builder instance, *and* returning values that are passed to later methods. So is the model here based on the methods mutating the builder to accumulate match operations into it, or is it based on the methods constructing an immutable value-type AST from the bottom up, or is it somehow both? (Maybe `builder` is a context that the AST nodes can be made in, but the relationship between those nodes is expressed by the return values? In that case, the initializer should probably be given some estimate information, much like `StringInterpolationProtocol`.)

2. If these methods are making AST nodes, should the method names indicate this? Perhaps by following the general `makeFoo` convention, or the result builder `buildFoo` convention? On the other hand, if they are having side effects on the builder, should they have verb-y names like `appendFoo` as we see in `StringInterpolationProtocol`?

3. Do we want to name metacharacters' methods based on specific semantics they're expected to follow? It might be better to have more mechanical translation rules, like:

> `\<identifier character>` turns into `makeEscaped_<identifier character>()` (except for characters that Swift string literals already assign a meaning to).

So for instance, the `__A2` line would be `let __A2 = builder.makeEscaped_w()`. That would free us from actually baking in a semantic for each character at this level, and would give us a consistent answer for characters that are used for different things in different dialects or aren't used in any of the dialects we know of. I also suspect it would be easier to remember than a lot of alternatives.

4. In `StringInterpolationProtocol`, I found that I didn't need a `finalize` because `init(stringInterpolation:)` could express any finalization logic I might need. The same might be true here.

Reply:
I do think these details @beccadax calls out are worth focusing on in this pitch. IIUC they're the protocol informal requirements and as such are more part of the literal than the concrete conforming `Regex` type we'll eventually provide.